### ISSUE TITLE
[BREAKING] Disallow integer arguments in select/transform/combine

### ISSUE THREAD

**Issue by matthieugomez** [JuliaData/DataFrames.jl#2436]
I was thinking it may be better to disallow integers as an argument in select/transform/combine, i.e. `select(df, 1)`.  

The main reason is that it creates an ambiguity between two different operations: (i) selecting the first column (ii) creating a column filled with ones.  The ambiguity arises in packages that make it easier to create the pair syntax from normal expressions (such as DataFramesMacros or DataFramesMetas). In these packages, the syntax `@select(df, 1)` could be about creating a vector of one or selecting the first column. 

Additionally, IMO, selecting using the name of a column is a better pattern than using its order. It makes sense to restrict integer indexing to `getindex`.

**Comment by bkamins** [comment]
> Additionally, IMO, selecting using the name of a column is a better pattern than using its order

Agreed, especially for retained code. However, I often see people writing in interactive sessions things like `combine(groupby(df, 1:3), 4:10 .=> mean)` just to save typing. Again - agreed that this is a bad habit, but I think we should also think about not killing the convenience if we do not have to.

> In these packages, the syntax `@select(df, 1)` could be about creating a vector of one or selecting the first column.

As you are aware, there is no abiguity in DataFrames.jl. Also I assume that in e.g. DataFramesMeta.jl `@select(df, 1)` would have a clear meaning of being one of these two operations. I understand that you are afraid that in `DataFrames.jl `select(df, 1)` selects the first column, while in DataFramesMeta.jl `@select(df, 1)` would create a column of `1`-s. But the question is if this is really plausible to happen (CC @pdeffebach) - as by writing `@select(df, 1)` you are not giving a column name. My understanding for the plan for `@select` in DataFramesMeta.jl was the following:
* leave all the syntax we currently have in DataFrames.jl allowed (if the user wants to use it)
* add an option to use `=` syntax

But then there is no ambiguity (unless I do not understand something), as for the "special rules" of DataFramesMeta.jl to kick-in you would have to use `=` in the expression you have written.

---

In summary - we could disallow integer indexing, as you can always write `names(df, 1)` instead of `1` in the places you used integer indexing in the past - but I do not currently see a strong justification for it (as it would be breaking, and I know this feature is used and is useful in certain scenarios (interactive use) - so we should have some strong rationale to change it).

**Comment by matthieugomez** [comment]
I’m not sure what the plan is. I was thinking DataFramesmacro would ultimately also transform expressions without LHS. @pdeffebach ?

In any case, we don’t know exactly how these things will work in the future. IMO, it is safer to have an overly restrictive syntax in 1.0, rather than a lax syntax that will make it harder to add syntax sugar in the future.

Btw, I like names(df, 1) — it fits very well with names(df, Int). It simples the API: use symbols to select columns. The function names(df...) makes it easy to convert type/integers into symbol.la.

**Comment by pdeffebach** [comment]
I agree with Bogumil. People refer to `select(df, 1)` all the time in interactive sessions. Given that it's a valid column identifier, `select(df, 1)` should return the first column, just like `df[!, [1]]`. 

With regards to DataFramesMeta, my plan is as Bogumil mentioned,

```
@select(df, 1)
```

would get parsed just like `select(df, 1)` because it doesn't start with a `x = ...` etc. 

My current plan is to have some sort of "flag" for non-LHS calls. 

```
@select(df, @fun 1) == select(df, [] => () -> 1)
```

This is clunky but I think would only be used by advanced users.

**Comment by matthieugomez** [comment]
Last thing: I agree with the convenience. But we can always improve the syntax post 1.0 to select columns by integers, types, or regex (which, for now, requires to use names(df,...)).

**Comment by matthieugomez** [comment]
@pdeffebach df.x works, but df.1 does not, so it's not like integers can be used everywhere symbols can.

In any case, I just wanted to make sure this issue was discussed before v1.0. Feel free to close if you think the status quo is better.

**Comment by bkamins** [comment]
The issue is:
1. I agree with you that `select(df, 1)` is not "clean"
2. But:
    a. people use it especially in interactive sessions
    b. removing it would be breaking, so we would get complaints from 2a.
3. It seems non conflicting with DataFramesMeta.jl

So I am closing it for now (still - thank you for raising this, as it always better to discuss things)

Now - `df.1` does not work, because it is an error on a parsing stage (you even could not intercept it in DataFramesMeta.jl). So there is even no way to discuss if we like it or not - simply we will not support it 😄.
